crates/xline/src/storage/index.rs

```diff
@@ -245,108 +378,250 @@ impl IndexOperate for Index {
+        match RangeType::get_range_type(key, range_end) {
+            RangeType::OneKey => {
+                Index::get_revision(&self.one_key_revisions(key, &self.state.lock()), revision)
+                    .map(|rev| vec![rev])
+                    .unwrap_or_default()
+            }
+            RangeType::AllKeys => self
+                .all_key_revisions()
+                .into_iter()
+                .filter_map(|(_, revs)| Index::get_revision(revs.as_ref(), revision))
+                .collect(),
+            RangeType::Range => self
+                .range_key_revisions(KeyRange::new(key, range_end))
+                .into_iter()
+                .filter_map(|(_, revs)| Index::get_revision(revs.as_ref(), revision))
+                .collect(),
+        }
+    }
+
+    fn register_revision(
         &self,
         key: Vec<u8>,
         revision: i64,
         sub_revision: i64,
-        create_revision: i64,
-        version: i64,
-    ) {
-        self.inner
-            .get_or_insert(key, RwLock::new(Vec::new()))
-            .value()
-            .map_write(|mut revisions| {
-                revisions.push(KeyRevision::new(
-                    create_revision,
-                    version,
+    ) -> (KeyRevision, Option<KeyRevision>) {
+        let index = &self.index_ref.inner;
+        let mut state = self.state.lock();
+
+        let next_rev = |revisions: &[KeyRevision]| {
+            let last = *revisions
+                .last()
+                .unwrap_or_else(|| unreachable!("empty revision list"));
+            let new_rev = if last.is_deleted() {
+                KeyRevision::new(revision, 1, revision, sub_revision)
+            } else {
+                KeyRevision::new(
+                    last.create_revision,
+                    last.version.overflow_add(1),
                     revision,
                     sub_revision,
-                ));
-            });
+                )
+            };
+            (new_rev, Some(last))
+        };
+
+        match (index.get(&key), state.entry(key)) {
+            (None, btree_map::Entry::Vacant(e)) => {
+                let new_rev = KeyRevision::new(revision, 1, revision, sub_revision);
+                let _ignore = e.insert(vec![new_rev]);
+                (new_rev, None)
+            }
+            (None | Some(_), btree_map::Entry::Occupied(mut e)) => {
+                let (new_rev, last) = next_rev(e.get_mut());
+                e.get_mut().push(new_rev);
+                (new_rev, last)
+            }
+            (Some(e), btree_map::Entry::Vacant(se)) => {
+                let (new_rev, last) = fmap_value(next_rev)(e);
+                let _ignore = se.insert(vec![new_rev]);
+                (new_rev, last)
+            }
+        }
     }
 
-    fn compact(&self, at_rev: i64) -> Vec<KeyRevision> {
-        let mut revs = Vec::new();
-        let mut del_keys = Vec::new();
+    fn prev_rev(&self, key: &[u8]) -> Option<KeyRevision> {
```

Comment:
When is this `prev_rev` invoked?

Reply:
It's a extra util to be used in the following refactor.

Reply:
Why do you name this method as `prev_rev`? Just consider the implementation of this method, the `current_rev` is a better name.

Reply:
renamed